### DISCUSSION TITLE
Support redirecting Android device audio

### DISCUSSION THREAD

**Discussion by madbrain76** [music-assistant/support#2471] [Feature requests and ideas]
Once upon a time, the Google Home app supported casting the Android device device audio to a group of Chromecast devices - or Chromecast audio devices in my case. Then, Sonos sued Google, and they removed this functionality from their app. It still supports redirecting the device audio (or screen), but only to a single Chromecast devices, and no longer to a group.

My main use case is to cast audio from a Youtube music video playing in Firefox with an ad blocker, and casting that to more than one room at a  time, say a whole floor, or the whole house. I'm aware that in this case, the audio and video are out of sync, but that's not a concern at all since I'm not even looking at the screen, and the phone may even be locked while I listen.

It should be noted that even using all-Google software (say, Youtube app, Youtube Music app, or Chrome, and the Google Home app) there is no way to accomplish this, even to a single Chromecast device, let alone a group, as far as I know. The Youtube app can only cast to Chromecast video devices, not Chromecast audio. This is probably due to A/V sync reasons. And the Youtube Music app can only cast audio from specific media designated as audio, not the audio from all Youtube videos, or my local device', or even my own Youtube channels video media..

I'm not sure of the technical requirements to implement this. It may require a rooted Android device (which I don' have, and don't really want to have for security reasons) or changes to the OS /APIs (having worked at Google, I won't hold my breath).  Hopefully I'm wrong about that requirement.

The only sure way I can think of to enable this functionality without root, which is admittedly a bit of a hack (ok, a lot) , is to take advantage of the "analog hole" of the Chromecast audio. This bypasses both the DRM, and group limitations. And there is no ARM. Well, not that kind of ARM, anyway.

It would work like this :
1. use the existing Google Home function to cast the device audio to a Chromecast audio device, that would only be used as a "sink" device. You could use a WiiM Pro/Plus device with better DACs and built-in CCA too, it just costs about 4-5x as much.
2. connect the analog output of the CCA to the analog input of an ADC supported by HA/MA . Not sure if there is such a thing yet
3. select the music provider audio input device in MA. Again, I don't think there is one yet, but I think I have seen an RFE for it.
4. select a group in MA (either a Chromecast group, or Universal)
5. play. 6. problem solved.
7. bonus point : use the optical out of the CCA in step 2, instead of analog. There is no DRM on it. And use an interface with SPDIF optical in on the HA/MA side. I have not used the CCA optical out myself. I have heard there are some limitations/bugs with it around sampling rates that don't apply to the analog out.

Obviously, there would be some pitfalls to this approach  : 
1. additional latency which further worsens the A/V sync, but also the delay in playback control (pause, etc)
2. the need to calibrate volume levels, unless one is using SPDIF
3. and audio degradation due the two conversions - DAC + ADC , unless one is using SPDIF.

Google Home device audio casting also some significant limitations.

1. if your phone roams to a different Wifi AP, the audio goes to hell (stutters), and never recovers until you stop and restart casting 
2. you can't cast device audio over Ethernet . I don't understand this one at all. All my WiiM devices, 1 CCA, both CCA w Google TV, are connected via Ethernet. But the Google Home app, if connected via a USB Ethernet NIC, doesn't even look for Chromecast devices.

There is also one major advantage to this approach - you can use it the same way with any audio or video device with analog audio components such as cassette decks, VCRs, turntables, reel-to-reel, Laserdisc, SACD, CD, DVD, Blu-ray, Or those with non-DRM digital outputs such as SPDIF. These could integrate very nicely with HA/MA if one has an IR transmitter, such as the 4 ESPHome AtHom AR-01 I use. Though it is very hard o impossible to maintain player state with IR-controlled devices as they are write-only.

**Comment by OzGav** [comment]
A new plugin has been added. Look out for Ariacast

**Comment by OzGav** [comment]
The feature request has been implemented successfully. As such this feature request will be closed and locked. If there are enhancements desired then open a new [feature request](https://github.com/orgs/music-assistant/discussions/categories/feature-requests-and-ideas). If there is a question about how this works then check the [documentation](https://beta.music-assistant.io/), or ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord. If there are problems then ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord or open an [issue](https://github.com/music-assistant/support/issues). Thanks for your interest in seeing this idea being brought to life.
